src/_hooks/useValidateSchema.ts

```diff
@@ -1,16 +1,43 @@
+import { isEqual } from 'lodash';
 import * as React from 'react';
 import * as yup from 'yup';
 
-export function useValidateSchema(schema?: yup.Schema<any>, value?: any, validateOpts?: yup.ValidateOptions): string[] {
-  return React.useMemo(() => {
-    if (!schema) return [];
+const noError: string[] = [];
+const unknownError = ['Input is invalid'];
 
-    try {
-      schema.validateSync(value, validateOpts);
-    } catch (e) {
-      return e.errors || ['Input is invalid'];
+export function useValidateSchema<T>(
+  schema?: yup.Schema<T>,
+  value?: T,
+  { abortEarly, recursive, context }: yup.ValidateOptions = {},
+): string[] {
+  const [errors, setErrors] = React.useState<string[]>(noError);
+  const memoizedSchema = useDeepCompareMemoize(schema);
+  const memoizedContext = useDeepCompareMemoize(context);
+
+  React.useEffect(() => {
+    if (!memoizedSchema) {
+      setErrors(noError);
+      return;
     }
+    memoizedSchema
+      .validate(value, { strict: true, abortEarly, recursive, context: memoizedContext }) // to avoid taking a useEffect dependency on validateOpts that is an object
+      .then(() => {
+        setErrors(noError);
+      })
+      .catch(e => {
+        setErrors(e.errors || unknownError);
+      });
+  }, [memoizedSchema, value, abortEarly, recursive, memoizedContext]);
+
+  return errors;
+}
+
+function useDeepCompareMemoize<T>(value: T) {
```

Comment:
Not too sure I understand why we need this; doesn't React.memo do the same thing?

Reply:
`React.memo` is shallow. In an ideal world, `schema` is memoized, but there are probably people who will call the component as:
```typescript
// in a component
const schema = yup.string().something();
return <FormInput schema={schema} />;
```
which is a new schema each time. We don't want to run the effect again in these cases, so a deep compare takes care of that.

Reply:
Hmm, I am not too sure the component should handle bad usages, I would prefer to document the behaviour and discipline the user. I do not know if this is a pattern or not but it's not blocking this though.

Reply:
In general I agree with you completely, but in this case I was very afraid that the above pattern is being used, and I cause a breaking change with this PR.

But I can take it out and see, should I?

Reply:
> In general I agree with you completely, but in this case I was very afraid that the above pattern is being used

Then we fix it in the codebase I'd say 😄

Reply:
I'm cool with that. I removed it!